File: ks/2021_F/4.cpp
#include <cstdio>
#include <map>

const int N = 20;
int n, m, k;
int l[N], r[N], a[N], edges[N];
std::map <int, int> cache;

int frontiers(int broken) {
    int ret = 0;
    for (int i = 0; i < n; i++) {
        if (broken & (1 << i)) {
            ret |= edges[i];
        }
    }
    return ret;
}

int recursive_count(long long points, int broken) {
    if (cache.count(broken)) {
        return cache[broken];
    }
    if (points > k) {
        return 0;
    }
    if (points == k) {
        return 1;
    }
    int ret = 0;
    int fs = frontiers(broken);
    for (int i = 0; i < n; i++) {
        if ((fs & (1 << i)) && ((broken & (1 << i)) == 0)) {
            if (l[i] <= points && points <= r[i]) {
                ret += recursive_count(points + a[i], broken | (1 << i));
            }
        }
    }
    cache[broken] = ret;
    return ret;
}

int main() {
    int T;
    scanf("%d", &T);
    for (int t = 0; t < T; t++) {
        scanf("%d%d%d", &n, &m, &k);
        cache.clear();
        for (int i = 0; i < n; i++) {
            edges[i] = 0;
            scanf("%d%d%d", &l[i], &r[i], &a[i]);
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            scanf("%d%d", &x, &y);
            edges[x] |= 1 << y;
            edges[y] |= 1 << x;
        }
        int ans = 0;
        for (int i = 0; i < n; i++) {
            ans += recursive_count(a[i], 1 << i);
        }
        printf("Case #%d: %d\n", t + 1, ans);
    }
    return 0;
}

File: ks/2021_F/4.py
from functools import lru_cache, reduce

if __name__ == "__main__":
    T = int(input())
    for case in range(T):
        n, m, k = tuple(map(int, input().strip().split()))
        rooms = []
        edges = [0] * n
        for _ in range(n):
            l, r, a = tuple(map(int, input().strip().split()))
            rooms.append((l, r, a))
        for _ in range(m):
            x, y = tuple(map(int, input().strip().split()))
            edges[x] |= 1 << y
            edges[y] |= 1 << x
        
        @lru_cache()
        def frontiers(broken):
            return reduce(lambda sa, sb: sa | sb, [edges[frontier] for frontier in range(n) if broken & (1 << frontier)])

        @lru_cache()
        def recursive_count(points: int, broken: int):
            if points == k:
                return 1
            ret = 0
            fs = frontiers(broken)
            for nxt in range(n):
                if (fs & (1 << nxt)) and (broken & (1 << nxt)) == 0:
                    if rooms[nxt][0] <= points and points <= rooms[nxt][1]:
                        nxt_tup = broken | (1 << nxt)
                        ret += recursive_count(points + rooms[nxt][2], nxt_tup)
            return ret
        
        ans = sum(recursive_count(rooms[i][2], 1 << i) for i in range(n))
        print(f"Case #{case + 1}: {ans}")
